store/memstore/memstore.go

```diff
@@ -0,0 +1,260 @@
+package memstore
+
+import (
+	"sync/atomic"
+
+	"cosmossdk.io/store/memstore/internal"
+	"cosmossdk.io/store/types"
+)
+
+var (
+	_ types.MemStoreManager = &memStoreManager{}
+	_ types.MemStore        = &memStore{}
+)
+
+type (
+	btree = internal.BTree
+
+	memStoreManager struct {
+		// root is an atomic pointer to the current root of the memStoreManager.
+		// When a branch is committed, it creates a new root node and atomically
+		// swaps it with the existing one.
+		root *atomic.Pointer[btree]
+		// The current B-tree is stored in the root atomic.Pointer when memStoreManager.Commit() occurs.
+		// The reason for creating it temporarily in this way is that it should not be read
+		// in the FinalizeBlock state before BaseApp.Commit happens.
+		//
+		// `current` is implemented with the assumption that it is accessed only by a single writer.
+		current *btree
+		// base ensures that only one branch (L1) can be committed.
+		//
+		// It is set to nil for Trees retrieved from the snapshotPool.
+		base *atomic.Pointer[btree]
+
+		snapshotPool SnapshotPool
+	}
+
+	// `memStore` implements a copy-on-write memStore operation pattern.
+	// Nested branches follow this structure as well, updating their parent's current on `Commit()`.
+	memStore struct {
+		// parent is nil for top-level branches, non-nil for nested branches
+		parent *memStore
+
+		// current holds the current working copy of the btree for this top-level branch.
+		current *btree
+
+		// base points to memStoreManager.base when the branch is created.
+		//
+		// If memstoreManager.base differs from base at commit time, the commit fails and a panic occurs.
+		base *btree
+
+		// manager is a reference to the parent memStoreManager, used by top-level branches to update the manager during commit.
+		manager *memStoreManager
+	}
+
+	UncommittableMemStore struct {
+		types.MemStore
+	}
+)
+
+func (u *UncommittableMemStore) Commit() {
+	panic("uncommittable MemStore cannot be committed")
```

Comment:
so this panics on commit but will happily pass on delete and set, is that acceptable? it's hard to reason about it.